packages/typespec-ts/test/modularUnit/serializeUtil.spec.ts

```diff
@@ -0,0 +1,1521 @@
+    const serializeUtil = await emitModularSerializeUtilsFromTypeSpec(
+      tspContent
+    );
+    assert.equal(serializeUtil?.length, 1);
+    assertEqualContent(
+      serializeUtil?.[0]?.getFullText()!,
+      `
+      import {
+        WidgetData1 as WidgetData1Rest,
+        WidgetData as WidgetDataRest,
+      } from "../rest/index.js";
+      import { WidgetData1, WidgetData } from "../models/models.js";
+      
+      /** serialize function for WidgetData1 */
+      function serializeWidgetData1(obj: WidgetData1): WidgetData1Rest {
+        return {
+          kind: obj["kind"],
+          start: obj["start"].toISOString(),
+          end: obj["end"]?.toISOString(),
+        };
+      }
+      
+      /** serialize function for WidgetData */
+      export function serializeWidgetData(obj: WidgetData): WidgetDataRest {
+        switch (obj.kind) {
+          case "kind1":
+            return serializeWidgetData1(obj);
+          default:
+            return obj;
+        }
+      }
+      `
+    );
+
+    const operationFiles = await emitModularOperationsFromTypeSpec(tspContent);
+    assert.ok(operationFiles);
+    assert.equal(operationFiles?.length, 1);
+    assertEqualContent(
+      operationFiles?.[0]?.getFullText()!,
+      `
+      import { TestingContext as Client } from "../rest/index.js";
+      import {
+        StreamableMethod,
+        operationOptionsToRequestParameters,
+        createRestError,
+      } from "@azure-rest/core-client";
+      export function _customGet1Send(
+        context: Client,
+        body: Widget1,
+        options: CustomGet1Options = { requestOptions: {} }
+      ): StreamableMethod<CustomGet1204Response> {
+        return context
+          .path("/customGet1")
+          .get({
+            ...operationOptionsToRequestParameters(options),
+            body: {
+              id: body["id"],
+              weight: body["weight"],
+              color: body["color"],
+              data: serializeWidgetData(body["data"]),
+            },
+          });
+      }
+      export async function _customGet1Deserialize(
+        result: CustomGet1204Response
+      ): Promise<void> {
+        if (result.status !== "204") {
+          throw createRestError(result);
+        }
+        return;
+      }
+      export async function customGet1(
+        context: Client,
+        body: Widget1,
+        options: CustomGet1Options = { requestOptions: {} }
+      ): Promise<void> {
+        const result = await _customGet1Send(context, body, options);
+        return _customGet1Deserialize(result);
+      }
+      `,
+      true
+    );
+  });
+
+  it("should generate serialize util if there's a special discriminated union variant of model with bytes property", async () => {
+    const tspContent = `
+    @discriminator("kind")
+    union WidgetData {
+      kind0: WidgetData0;
+      kind1: WidgetData1;
+    }
+
+    model WidgetData0 {
```

Comment:
It would be good to include more test cases like:

WidgetData0 & WidgetData1 all contains the data of binary.
WidgetData0 contains the data of datetime and WidgetData1 contains the data of binary.

Reply:
For case 1, probably not, for case 2, I think I have a case. let me find the link.